### PR TITLE
PP-35: Failover file locks are broken if file timestamps don't move for a long enough period (STONITH)

### PR DESCRIPTION
**Pull request by sid2364**
#### Issue-ID

* **[PP-35](https://pbspro.atlassian.net/browse/PP-35)**

#### Problem description
* Split brain scenario was not impossible and could occur due to the fact that NFS is used both as a datastore as well as a quorum server.

#### Cause / Analysis
* Primary could still be active when secondary decides to take over, and would result in two data services running and possibly corrupting the database.

#### Solution description
* A STONITH (Shoot The Other Node In The Head) script could be used to bring down the primary node if it is alive, to ensure that when secondary comes up, primary is definitely not up.

#### Links
* [Community Forum Discussion](http://community.pbspro.org/t/pp-35-pp-729-pbs-failover-stonith/658)
* [Design Document](https://pbspro.atlassian.net/wiki/spaces/PD/pages/63766571/PP-35+PP-729+PBS+Failover+-+STONITH)

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [x] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [x] I have added documentation in the [project documentation area](https://pbspro.atlassian.net/wiki/spaces/PD/pages/63766571/PP-35+PP-729+PBS+Failover+-+STONITH).
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [x] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

**Comment by sid2364** [comment]
Adding a note saying the design document is located [here](https://pbspro.atlassian.net/wiki/spaces/PD/pages/63766571/PP-35+PP-729+PBS+Failover+-+STONITH).

**Comment by nithinj** [comment]
I could not find any test logs or manual test steps in the ticket even though you've checked both the items in the checklist. Please make sure that you test for both Windows and Linux as the code contains changes for both.

**Comment by sid2364** [comment]
@nithinj Have added the test logs for Linux.

File: src/server/failover.c [hunk 1]
```diff
@@ -806,6 +810,142 @@ alt_conn(pbs_net_t addr, unsigned int sec)
...
+	char		stonith_cmd[3*MAXPATHLEN+1] = {0};
+	char		stonith_fl[MAXPATHLEN+1] = {0};
+	char		*p = NULL;
+	char		out_err_fl[MAXPATHLEN+1] = {0};
+	char		*out_err_msg = NULL;
+	int		rc = 0;
+	int		fd = 0;
+	struct stat	stbuf;
+#ifdef WIN32
+	char buf[MAXPATHLEN+1] = {0};
```
Comment (nithinj) [inline src/server/failover.c]: You can go for a more meaningful name if `buf` is used to hold just the stonith script location.

Comment (sid2364) [reply src/server/failover.c]: Have changed it.

File: src/server/failover.c [hunk 2]
```diff
@@ -806,6 +810,142 @@ alt_conn(pbs_net_t addr, unsigned int sec)
...
+#endif /* WIN32 */
+	if (node == NULL )
+		return -1;
+
+#ifdef WIN32
```
[the remark this answers is not in the document]
Comment (sid2364) [reply src/server/failover.c]: Yes, makes sense. I've removed the buf_pbs_home_path variable, also changed the way I was building the error file path.

File: src/server/failover.c [hunk 3]
```diff
@@ -806,6 +810,142 @@ alt_conn(pbs_net_t addr, unsigned int sec)
...
+			p = out_err_msg + strlen(out_err_msg) - 1;
+
+			while ((p >= out_err_msg) && (*p == '\r' || *p == '\n'))
+				*p-- = 0; /* supress the last newline */
+		}
+		close(fd);
+	}
+
+	if (out_err_msg) {
+		snprintf(log_buffer, LOG_BUF_SIZE,
```
[the remark this answers is not in the document]
Comment (sid2364) [reply src/server/failover.c]: The log message would get cut off at LOG_BUF_SIZE. In any case, LOG_BUF_SIZE is 4096, which should be enough for getting a good idea about the actual error.

File: src/server/failover.c [hunk 4]
```diff
@@ -806,6 +810,132 @@ alt_conn(pbs_net_t addr, unsigned int sec)
...
+		}
+	}
+
+	/* create unique filename by appending pid */
+	snprintf(out_err_fl, sizeof(out_err_fl), "%s/spool/stonith_out_err_fl_%s_%d", pbs_conf.pbs_home_path, node, getpid());
+	out_err_fl[sizeof(out_err_fl) - 1] = '\0';
+#ifdef WIN32
+	repl_slash(out_err_fl);
+
+	/* execute stonith script and redirect output to file */
```
Comment (nithinj) [inline src/server/failover.c]: Looks like we still have some redundant code for both platforms which can be optimized.

Comment (sid2364) [reply src/server/failover.c]: Have updated it.

File: src/server/failover.c [hunk 5]
```diff
@@ -1138,6 +1266,17 @@ be_secondary(time_t delay)
 					}
 					break;
 				}
+				/* Invoke stonith, to make sure primary is down */
+				rc = check_and_invoke_stonith(pbs_conf.pbs_primary);
+				if (rc) {
+					Secondary_state = SECONDARY_STATE_takeov;
```
Comment (bhroam) [inline src/server/failover.c]: Won't Secondary_state be SECONDARY_STATE_takeov?  You are in that case of the switch() statement.

File: src/server/failover.c [hunk 6]
```diff
@@ -1138,6 +1266,17 @@ be_secondary(time_t delay)
...
+					sprintf(log_buffer, "Secondary will attempt taking over again");
```
Comment (bhroam) [inline src/server/failover.c]: Use snprintf()

File: src/server/failover.c [hunk 7]
```diff
@@ -806,6 +810,129 @@ alt_conn(pbs_net_t addr, unsigned int sec)
 
 /**
  * @brief
+ * 	Function to check if stonith script exists at PBS_HOME/server_priv/stonith.
+ * 	If it does then invoke the script for execution.
+ * 
+ * @param[in]	node - hostname of the node, that needs to be powered off.
```
[the remark this answers is not in the document]
Comment (sid2364) [reply src/server/failover.c]: I've changed it to "that needs to be brought down".

File: src/server/failover.c [hunk 8]
```diff
@@ -806,6 +810,129 @@ alt_conn(pbs_net_t addr, unsigned int sec)
...
+			log_buffer[LOG_BUF_SIZE - 1] = '\0';
+			log_event(PBSEVENT_SYSTEM, PBS_EVENTCLASS_SERVER, LOG_INFO,
+				  msg_daemonname, log_buffer);
+			return 0;
+		}
+	}
+
+	/* create unique filename by appending pid */
+	snprintf(out_err_fl, sizeof(out_err_fl), "%s/spool/stonith_out_err_fl_%s_%d", pbs_conf.pbs_home_path, node, getpid());
+	out_err_fl[sizeof(out_err_fl) - 1] = '\0';
```
[the remark this answers is not in the document]
Comment (sid2364) [reply src/server/failover.c]: Have changed it throughout.

File: src/server/failover.c [hunk 9]
```diff
@@ -806,6 +810,129 @@ alt_conn(pbs_net_t addr, unsigned int sec)
...
+	
+#ifdef WIN32
+	repl_slash(out_err_fl);
+#endif /* WIN32 */
+
+	/* execute stonith script and redirect output to file */
+	snprintf(stonith_cmd, sizeof(stonith_cmd), "%s %s > %s 2>&1", stonith_fl, node, out_err_fl);
```
[the remark this answers is not in the document]
Comment (sid2364) [reply src/server/failover.c]: Yes, the redirect does work. [Source](https://support.microsoft.com/en-us/help/110930/redirecting-error-messages-from-command-prompt-stderr-stdout).

File: src/server/failover.c [hunk 10]
```diff
@@ -806,6 +810,129 @@ alt_conn(pbs_net_t addr, unsigned int sec)
...
+	log_buffer[LOG_BUF_SIZE - 1] = '\0';
+	log_event(PBSEVENT_SYSTEM, PBS_EVENTCLASS_SERVER, LOG_NOTICE,
+			msg_daemonname, log_buffer);
+
+#ifdef WIN32
+	rc = wsystem(stonith_cmd, INVALID_HANDLE_VALUE);
+#else 
+	rc = system(stonith_cmd);
+	if (WIFEXITED(rc))
+		rc = WEXITSTATUS(rc);
```
[the remark this answers is not in the document]
Comment (sid2364) [reply src/server/failover.c]: Changed.

File: src/server/failover.c (hunk 5 above)
Comment (sid2364) [reply src/server/failover.c]: Was doing this to make sure it comes back to try to execute the script again if an error occurs, but the Secondary_state shouldn't change anywhere in this block except when connection with primary is re-established.

Have removed it.

**Comment by bhroam** [comment]
Looks good to me.  I sign off.

**Comment by bhroam** [comment]
Once Nithin signs off again, squash and rebase for the final merge.

**Comment by nithinj** [comment]
Looks good to me. I sign-off.

**Comment by sid2364** [comment]
Thanks @bhroam and @nithinj for reviewing this. I've squashed the commits and rebased the branch.

**Comment by bhroam** [comment]
Something got merged in before you.  You'll need to rebase again.  It's probably best if @agrawalravi90 merges this since he's in your timezone.

**Pull request merged**